ERCS/erc-7677.md

```diff
@@ -0,0 +1,350 @@
+
+This method MAY return paymaster-specific gas values if applicable to the provided EntryPoint version. For example, if provided with EntryPoint v0.7, this method MAY return `paymasterVerificationGasLimit` and `paymasterPostOpGasLimit` values. The wallet SHOULD use these provided gas values when submitting the `UserOperation` to a bundler for gas estimation.
+
+##### `pm_getPaymasterStubData` RPC Specification
+
+Note that the user operation parameter does not include a signature, as the user signs after all other fields are populated.
+
+```typescript
+// [userOp, entryPoint, chainId, context]
+type GetPaymasterStubDataParams = [
+  // Below is specific to Entrypoint v0.6 but this API can be used with other entrypoint versions too
+  {
+    sender: `0x${string}`;
+    nonce: `0x${string}`;
+    initCode: `0x${string}`;
+    callData: `0x${string}`;
+    callGasLimit: `0x${string}`;
+    verificationGasLimit: `0x${string}`;
+    preVerificationGas: `0x${string}`;
+    maxFeePerGas: `0x${string}`;
+    maxPriorityFeePerGas: `0x${string}`;
+  }, // userOp
+  `0x${string}`, // EntryPoint
+  `0x${string}`, // Chain ID
+  Record<string, any> // Context
+];
+
+type GetPaymasterStubDataResult = {
+  paymaster?: string // Paymaster address (entrypoint v0.7)
+  paymasterData?: string; // Paymaster data (entrypoint v0.7)
+  paymasterVerificationGasLimit?: `0x${string}`; // Paymaster validation gas (entrypoint v0.7)
+  paymasterPostOpGasLimit?: `0x${string}`; // Paymaster post-op gas (entrypoint v0.7)
+  paymasterAndData?: string; // Paymaster and data (entrypoint v0.6)
+}
+```
+
+###### `pm_getPaymasterStubData` Example Parameters
+
+```json
+[
+  {
+    "sender": "0x...",
+    "nonce": "0x...",
+    "initCode": "0x",
+    "callData": "0x...",
+    "callGasLimit": "0x...",
+    "verificationGasLimit": "0x...",
+    "preVerificationGas": "0x...",
+    "maxFeePerGas": "0x...",
+    "maxPriorityFeePerGas": "0x...",
+  },
+  "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
+  "0x2105",
+  {
+    // Illustrative context field. These should be defined by service providers.
+    "policyId": "962b252c-a726-4a37-8d86-333ce0a07299"
+  }
+]
+```
+
+###### `pm_getPaymasterStubData` Example Return Value
+
+Paymaster services MUST detect which entrypoint version the account is using and return the correct fields.
+
+For example, if using entrypoint v0.6:
+
+```json
+{
+  "paymasterAndData": "0x..."
+}
+```
+
+If using entrypoint v0.7:
+
+```json
+{
+  "paymaster": "0x...",
+  "paymasterData": "0x..."
+}
+```
+
+If using entrypoint v0.7, with paymaster gas estimates:
+
+```json
+{
+  "paymaster": "0x...",
+  "paymasterData": "0x...",
+  "paymasterVerificationGasLimit": "0x...",
+  "paymasterPostOpGasLimit": "0x..."
```

Comment:
in v0.7, postOpGasLimit **cannot** be estimated by app/wallet, and must be returned by the serve
verificationGasLimit MAY be estimated, though if the service knows the value, it can place it anyway.